File: python/fastapi/src/app/modules/transactions/service.py
from typing import List
from uuid import UUID

# Common
from app.common.helpers import has_permission
from app.common.enums import RoleEnum
from app.common.errors.either import Either, Left, Right
from app.common.errors.base_error import BaseError
from app.common.errors import BadRequestError, NotFoundError, UnauthorizedError, InternalServerError
from app.common.messages import MESSAGES
from app.database import db

# Transactions
from .entity import TransactionEntity
from .enums.transaction_type import TransactionTypeEnum

# DTOs
from .dtos.inputs.create_transaction import CreateTransactionDto

class TransactionsService:
    def __init__(self, repositories):
        self._repositories = repositories

    def create(self, data: CreateTransactionDto, id_client: str) -> Either[BaseError, TransactionEntity]:
        tipo = TransactionTypeEnum(data.tipo)

        origin_account = None

        if data.idContaDestino == data.idContaOrigem:
            return Left(BadRequestError(MESSAGES['error']['account']['BadRequest']['SameAccount']))

        # Verifica se contas existem
        if data.idContaOrigem:
            origin_account = self._repositories['accounts'].find_by_id(data.idContaOrigem)

            if not origin_account:
                return Left(NotFoundError(MESSAGES['error']['account']['NotFoundOrigin']))

            if origin_account['idCliente'] != UUID(id_client):
                return Left(UnauthorizedError())

        if data.idContaDestino:
            target_account = self._repositories['accounts'].find_by_id(data.idContaDestino)

            if not target_account:
                return Left(NotFoundError(MESSAGES['error']['account']['NotFoundTarget']))

        if (tipo == TransactionTypeEnum.TRANSFER or tipo == TransactionTypeEnum.WITHDRAWAL) and origin_account['saldo'] < data.valor:
            return Left(BadRequestError(MESSAGES['error']['account']['BadRequest']['BalanceNotEnough']))

        with db.connect() as conn:
            trx = conn.begin()

            try:
                if tipo == TransactionTypeEnum.DEPOSIT:
                    self._repositories['accounts'].add_balance(data.idContaDestino, data.valor, conn)
                elif tipo == TransactionTypeEnum.TRANSFER:
                    self._repositories['accounts'].remove_balance(data.idContaOrigem, data.valor, conn)
                    self._repositories['accounts'].add_balance(data.idContaDestino, data.valor, conn)
                elif tipo == TransactionTypeEnum.WITHDRAWAL:
                    self._repositories['accounts'].remove_balance(data.idContaOrigem, data.valor, conn)

                new_transaction = self._repositories['transactions'].create({
                    **{**data.model_dump(), 'tipo': tipo.value}
                }, conn)

                trx.commit()

                return Right(new_transaction)
            except Exception as e:
                trx.rollback()
                return Left(InternalServerError())          

        

    def find_all(self, id_account: str, id_client: str, role: RoleEnum) -> Either[BaseError, List[TransactionEntity]]:
        permission = has_permission(role, RoleEnum.MANAGER)

        account = self._repositories['accounts'].find_by_id(id_account)

        if not account:
            return Left(NotFoundError(MESSAGES['error']['account']['NotFound']))

        if account['idCliente'] != UUID(id_client) and not permission:
            return Left(UnauthorizedError())

        transactions = self._repositories['transactions'].find_all(id_account)

        return Right(transactions)

File: go/fiber_orm/internal/modules/transaction/service.go
package transaction

import (
	r "tcc/internal/common/enum"
	"tcc/internal/common/errors"
	"tcc/internal/common/helper"
	"tcc/internal/common/messages"
	"tcc/internal/database"
	repository "tcc/internal/modules/account"
	account "tcc/internal/modules/account/entity"
	"tcc/internal/modules/transaction/dto"
	"tcc/internal/modules/transaction/entity"
	"tcc/internal/modules/transaction/enum"

	"github.com/google/uuid"
)

type TransactionService struct{
	repository struct{
		account repository.AccountRepository
		transaction TransactionsRepository
	}
}

func (s *TransactionService) Create(data dto.CreateTransactionDto, idClient string) (*entity.TransactionEntity, *errors.BaseError) {
	
	var tipo = data.Tipo
	var originAccount *account.AccountEntity = nil

	if data.IDContaDestino == data.IDContaOrigem {
		return nil, errors.BadRequestError(messages.ErrorMessages.Account.BadRequest.SameAccount)
	}

	
	// Verificar se contas existem
	if data.IDContaOrigem != nil {
		originAccount, _ = s.repository.account.FindById(*data.IDContaOrigem, false)
	
		if originAccount == nil {
			return nil, errors.NotFoundError(messages.ErrorMessages.Account.NotFoundOrigin)
		}

		if originAccount.IDCliente.String() != idClient {
			return nil, errors.UnauthorizedError(messages.ErrorMessages.Unauthorized)
		}
	}

	if data.IDContaDestino != nil {
		targetAccount, _ := s.repository.account.FindById(*data.IDContaDestino, false)

		if targetAccount == nil {
			return nil, errors.NotFoundError(messages.ErrorMessages.Account.NotFoundTarget)
		}
	}


	if (tipo == enum.TRANSFER || tipo == enum.WITHDRAWAL) && (originAccount != nil && *originAccount.Saldo < data.Valor) {
		return nil, errors.BadRequestError(messages.ErrorMessages.Account.BadRequest.BalanceNotEnough)
	}


	// Iniciar transação
	tx := database.Conn.Begin();

	switch tipo {
		case enum.DEPOSIT:
			s.repository.account.AddBalance(*data.IDContaDestino, data.Valor, tx)
		case enum.TRANSFER:
			s.repository.account.RemoveBalance(*data.IDContaOrigem, data.Valor, tx)
			s.repository.account.AddBalance(*data.IDContaDestino, data.Valor, tx)
		case enum.WITHDRAWAL:
			s.repository.account.RemoveBalance(*data.IDContaOrigem, data.Valor, tx)
	}

	newTransaction, err := s.repository.transaction.Create(ICreateTransactionData{
		Tipo: tipo,
		Valor: data.Valor,
		IDContaOrigem: func() *uuid.UUID {
			if data.IDContaOrigem != nil {
				value := uuid.MustParse(*data.IDContaOrigem)
				return &value
			}
			return nil
		}(),
		IDContaDestino: func() *uuid.UUID {
			if data.IDContaDestino != nil {
				value := uuid.MustParse(*data.IDContaDestino)
				return &value
			}
			return nil
		}(),
	}, tx)

	if err != nil {
		tx.Rollback()
		return nil, errors.InternalServerError(messages.ErrorMessages.InternalServer)
	}

	err = tx.Commit().Error
	if err != nil {
		return nil, errors.InternalServerError(messages.ErrorMessages.InternalServer)
	}

	return newTransaction, nil
}

func (s *TransactionService) FindAll(idAccount string, idClient string, role r.RoleEnum) ([]*entity.TransactionEntity, *errors.BaseError) {
	permission := helper.HasPermission(role, r.MANAGER)

	account, _ := s.repository.account.FindById(idAccount, false)

	if account == nil {
		return nil, errors.NotFoundError(messages.ErrorMessages.Account.NotFound)
	}

	if account.IDCliente.String() != idClient && !permission {
		return nil, errors.UnauthorizedError(messages.ErrorMessages.Unauthorized)
	}

	transactions, _ := s.repository.transaction.FindAll(idAccount)

	if transactions == nil {
		return []*entity.TransactionEntity{}, nil
	}

	return transactions, nil
}